config/src/main/java/com/quorum/tessera/config/adapters/KeyDataAdapter.java

```diff
@@ -27,12 +26,44 @@ public ConfigKeyPair unmarshal(final KeyData keyData) {
         }
 
         //case 3, the keys are provided inside a file
-        return new FilesystemKeyPair(keyData.getPublicKeyPath(), keyData.getPrivateKeyPath());
+        if(keyData.getPublicKeyPath() != null && keyData.getPrivateKeyPath() != null) {
+            return new FilesystemKeyPair(keyData.getPublicKeyPath(), keyData.getPrivateKeyPath());
+        }
+
+        //case 4, the key config specified is invalid
+        return new UnsupportedKeyPair(
+            keyData.getConfig(),
+            keyData.getPrivateKey(),
+            keyData.getPublicKey(),
+            keyData.getPrivateKeyPath(),
+            keyData.getPublicKeyPath()
+        );
     }
 
     @Override
     public KeyData marshal(final ConfigKeyPair keyData) {
-        return keyData.marshal();
-    }
+        ConfigKeyPairType type = keyData.getType();
 
+        if(type == DIRECT) {
+            DirectKeyPair kp = (DirectKeyPair) keyData;
+            return new KeyData(null, kp.getPrivateKey(), kp.getPublicKey(), null, null);
+        }
+
+        if(type == INLINE) {
+            InlineKeypair kp = (InlineKeypair) keyData;
+            return new KeyData(kp.getPrivateKeyConfig(), null, kp.getPublicKey(), null, null);
+        }
+
+        if(type == FILESYSTEM) {
+            FilesystemKeyPair kp = (FilesystemKeyPair) keyData;
+            return new KeyData(null, null, null, kp.getPrivateKeyPath(), kp.getPublicKeyPath());
+        }
+
+        if(type == UNSUPPORTED) {
+            UnsupportedKeyPair kp = (UnsupportedKeyPair) keyData;
+            return new KeyData(kp.getConfig(), kp.getPrivateKey(), kp.getPublicKey(), kp.getPrivateKeyPath(), kp.getPublicKeyPath());
+        }
+
+        throw new RuntimeException("The keypair type " + keyData.getClass() + " is not allowed");
```

Comment:
Can we make this and illegal arguement, unsupported operation exception or some custom exception. We need to have a test for this scenario and we don't want to assert on a RuntimeException as null pointer will give you a positive outcome